packages/flutter/test/widgets/editable_text_test.dart

```diff
@@ -9514,6 +9514,452 @@ void main() {
       isNot(contains(matchesMethodCall('TextInput.requestAutofill'))),
     );
   });
+
+  group('UndoStack', () {
```

Comment:
I understand that `UndoRedo` is split for code modularity and is meant to be private but testable. But can we just test `EditableText` in all of the following tests? The behavior of `EditableText` is what the users care and we maintain backward compatibility of. By making `UndoRedo` a private class, you're saying you'd like to refactored or broken in the future, at which point of these tests will have to be migrated.

Reply:
I usually write small building blocks like this and then test them to convince myself that they work the way that I expect them to.  It's useful for me personally to have the tests even if we refactor or change behavior.  I don't want to burden everyone else if that's not Flutter's style though...

Reply:
The problem is mostly for `visibleForTesting` since it's not enforced by the compiler but by the linter, which the developer can turn off. The last time (which is a long time ago) I was told that it was discouraged (but not actually banned.) But I do think it reveals a problem where you're testing something that the users won't see, so might not be kept compatible during future development at all, i.e. adding unnecessary compatibility requirements. Anyway, just a soft suggestion.

Reply:
Ok thanks for the explanation.  I probably am going against Flutter trends a little bit.  I'll keep an eye out for similar patterns and problems with this code and consider changing it then.  It wouldn't be terrible to just test the widget.

Reply:
+1 to what @dkwingsmt has said, it feels like these tests are testing implementation detail.

Reply:
Alright I'm outnumbered, I'll remove the tests and make the class fully private.  I'll also add another test to TextEditingHistory to make sure all of those cases are covered.